src/runtime_src/core/pcie/driver/linux/xocl/subdev/xgq.c

```diff
@@ -659,19 +741,105 @@ static int xgq_load_xclbin(struct platform_device *pdev,
 	u64 xclbin_len = xclbin->m_header.m_length;
 	int ret = 0;
 	
-	if (down_interruptible(&xgq->xgq_data_sema)) {
-		XGQ_ERR(xgq, "XGQ data transfer is interrupted");
-		return -EIO;
-	}
-
 	ret = xgq_transfer_data(xgq, u_xclbin, xclbin_len,
 		XGQ_CMD_OP_LOAD_XCLBIN, XOCL_XGQ_DOWNLOAD_TIME);
 
-	up(&xgq->xgq_data_sema);
-
 	return ret == xclbin_len ? 0 : -EIO;
 }
 
+static int xgq_log_page_fw(struct platform_device *pdev,
+	char **fw, size_t *fw_size)
+{
+	struct xocl_xgq_vmr *xgq = platform_get_drvdata(pdev);
+	struct xocl_xgq_vmr_cmd *cmd = NULL;
+	struct xgq_cmd_log_payload *payload = NULL;
+	struct xgq_cmd_sq_hdr *hdr = NULL;
+	int ret = 0;
+	int id = 0;
+	u32 address = 0;
+	u32 len = LOG_PAGE_SIZE;
+
+	cmd = kzalloc(sizeof(*cmd), GFP_KERNEL);
+	if (!cmd) {
+		XGQ_ERR(xgq, "kmalloc failed, retry");
+		return 0;
+	}
+
+	cmd->xgq_cmd_cb = xgq_complete_cb;
+	cmd->xgq_cmd_arg = cmd;
+	cmd->xgq_vmr = xgq;
+
+	if (shm_acquire_log_page(xgq, &address)) {
+		ret = -EIO;
+		goto done;
+	}
+
+	payload = &(cmd->xgq_cmd_entry.log_payload);
+	payload->address = address;
+	payload->size = len;
+	payload->offset = 0;
+	payload->pid = XGQ_CMD_LOG_FW;
+
+	hdr = &(cmd->xgq_cmd_entry.hdr);
+	hdr->opcode = XGQ_CMD_OP_GET_LOG_PAGE;
+	hdr->state = XGQ_SQ_CMD_NEW;
+	hdr->count = sizeof(*payload);
+	id = get_xgq_cid(xgq);
+	if (id < 0) {
+		XGQ_ERR(xgq, "alloc cid failed: %d", id);
+		goto done;
+	}
+	hdr->cid = id;
+
+	/* init condition veriable */
+	init_completion(&cmd->xgq_cmd_complete);
+
+	/* set timout actual jiffies */
+	cmd->xgq_cmd_timeout_jiffies = jiffies + XOCL_XGQ_CONFIG_TIME;
+
+	ret = submit_cmd(xgq, cmd);
+	if (ret) {
+		XGQ_ERR(xgq, "submit cmd failed, cid %d", id);
+		/* return 0, because it is not a firewall trip */
+		ret = 0;
+		goto done;
+	}
+
+	/* wait for command completion */
+	wait_for_completion_interruptible(&cmd->xgq_cmd_complete);
+
+	ret = cmd->xgq_cmd_rcode;
+
+	if (ret) {
+		XGQ_ERR(xgq, "ret %d", ret);
+	} else {
+		struct xgq_cmd_cq_log_page_payload *fw_result = NULL;
+
+		vmr_cq_result_copy(xgq, cmd);
+		fw_result = (struct xgq_cmd_cq_log_page_payload *)&xgq->xgq_cq_payload;
+
+		if (fw_result->count > len) {
+			XGQ_ERR(xgq, "need to alloc %d for device data", 
+				fw_result->count);
+			ret = -ENOSPC;
+		} else {
+			*fw_size = fw_result->count;
+			*fw = vmalloc(*fw_size);
+			memcpy_from_device(xgq, address, *fw, *fw_size);
+			ret = 0;
+			XGQ_INFO(xgq, "loading fw from vmr size %ld", *fw_size);
+		}
+	}
+done:
+	shm_release_log_page(xgq);
+
+	if (cmd) {
+		remove_xgq_cid(xgq, id);
```

Comment:
id will be zero if shm_acquire_log_page() fails. Will it be ok to free id 0?

Reply:
Yes, I added checking for this. We could refactor the code after this release and move "cmd allocation, cid allocation and shared memory reservation" into one function. But for this release, we may keep it as is for not having potential regression.